chainservice/builder.go

```diff
@@ -385,9 +384,23 @@ func (builder *Builder) buildNodeInfoManager() error {
 	if stk == nil {
 		return errors.New("cannot find staking protocol")
 	}
-	dm := nodeinfo.NewInfoManager(&builder.cfg.NodeInfo, cs.p2pAgent, cs.chain, builder.cfg.Chain.ProducerPrivateKey(), func(ctx context.Context) (state.CandidateList, error) {
-		return stk.ActiveCandidates(ctx, cs.factory, 0)
+
+	dm := nodeinfo.NewInfoManager(&builder.cfg.NodeInfo, cs.p2pAgent, builder.cfg.Chain.ProducerPrivateKey(), func() []string {
+		whiteList := []string{}
+		candidates, err := stk.ActiveCandidates(context.Background(), cs.factory, 0)
+		if err != nil {
+			log.L().Error("failed to get active candidates", zap.Error(errors.WithStack(err)))
+		} else {
+			whiteList = make([]string, len(candidates))
```

Comment:
`whitelist` already defined, and is empty []string at L389, why make again? i think you want to allocate the exact size, then this maybe be better?
```
if err != nil {
	log.L().xxx
	return nil
}

whiteList = make([]string, len(candidates))
...
return whiteList
```